### PR TITLE
Improve extract_authors script to reduce time complexity

### LINKED ISSUE #29882: Optimize `make update-authors`
One of the steps during the release process is to ensure that the list of `AUTHORS` in the repo is up to date. We currently manage this by running `make update-authors`. This currently takes me about 3.5 minutes on my system to run. It'd be nice if this ran within a few seconds.

### PR DESCRIPTION
**Pull request by shivam-Purohit**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

The current command `make update-authors ` that uses the script `extract_authors.sh` takes about 3 and half minutes to complete.  This change focuses on reducing the overall time and improving the script results

Fixes: #29882

### CODE REVIEW

**Comment by shivam-Purohit** [comment]
@20RitikSingh had a nice way of solving this problem but it had few issues that were needed to address. I think its inactive now so I thought of making changes myself
1. to ignore the `vagrant` keyword
2.  Is as @bimmlerd suggested the `Alexandre Perrin `was repeated twice in the git log output
the reason for this behaviour was the uniq command was running for the whole line
for example
```
Alexandre Perrin                        alex@isovalent.com
Alexandre Perrin                        alex@kaworu.ch
```

alex has two mails associated with him so it considers it as a whole new line rather than a duplicate
similar cases occur in other authors as well.
the way to solve it is to just check for the first 40 characters ( that is used to separate name from email)


also there are other anomalies like the repeating name for
```
adam.bocim                              adam.bocim@firma.seznam.cz
Adam Bocim                              adam.bocim@seznam.cz
```

i was confused if to consider them as different or same authors because the command below does give them as different authors
```
git shortlog --summary \
		  | awk '{$1=""; print $0}' \
		  | sed -e 's/^ //' \
			-e '/vagrant/d'
```

with both the entires
we use to ignore the by using head -1 while using
`git log --use-mailmap --author="$i" --format="%<|(40)%aN%aE" | head -1`

which otherwhile would output something like this for author

```
vboxuser@shivam~Desktoplfxcilium$ git log --use-mailmap --author=Adam Bocim --format=%(40)%aN%aE
Adam Bocim                              adam.bocim@seznam.cz
Adam Bocim                              adam.bocim@seznam.cz
Adam Bocim                              adam.bocim@seznam.cz
```

```
vboxuser@shivam~Desktoplfxcilium$ git log --use-mailmap --author=adam.bocim --format=%(40)%aN%aE
Adam Bocim                              adam.bocim@seznam.cz
Adam Bocim                              adam.bocim@seznam.cz
Adam Bocim                              adam.bocim@seznam.cz
adam.bocim                              adam.bocim@firma.seznam.cz
```

which is very confusing.
Other anomalies are

```
vboxuser@shivam:~/Desktop/lfx/cilium$ git log --author="Li Cheng" --format="%<|(40)%aN%aE"
Li Chengyuan                            chengyuanli@hotmail.com
Li Chengyuan                            chengyuanli@hotmail.com
Li Chengyuan                            chengyli@ebay.com
Li Cheng                                rapid.li@huolala.cn
Li Cheng                                rapid.li@huolala.cn
Li Cheng                                rapid.li@huolala.cn


vboxuser@shivam:~/Desktop/lfx/cilium$ git log --author="Li Chengyuan" --format="%<|(40)%aN%aE"
Li Chengyuan                            chengyuanli@hotmail.com
Li Chengyuan                            chengyuanli@hotmail.com
Li Chengyuan                            chengyli@ebay.com
```

the changes also introduces some other users that were ignored before which can be found by comparing both the files

**Review by bimmlerd — CHANGES_REQUESTED** [review]
Hi, thanks for your contribution!

So I think the proper way to solve all of the above problems is to use https://git-scm.com/docs/gitmailmap. We have a `.mailmap` file already, but it's not complete. The idea of comparing the names is okay, but it does not solve the problem for other tools in the future, the way the mailmap can.

Here's an example of what I'd consider the proper fix. Let's look at "André Martins":

André has an entry in the mailmap already, it maps his canonical email (andre@cilium.io) to his name. This is here to to fix the broken name in commit 4206a8cc9. But he has used other emails to author commits, which didn't show up in the existing script (thanks to `head -1`). Specifically, the commit log contains (`git log --use-mailmap --format="%<|(40)%aN%aE" | sort -fu | grep Martins`):

```
André Martins                           aanm90@gmail.com
André Martins                           aanm@users.noreply.github.com
André Martins                           andre@cilium.dev
```

Let's write `.mailmap` entries to fix those emails - the complete mailmap for him would then be:

```
André Martins <andre@cilium.io>
<andre@cilium.io> <andre@cilium.dev>
<andre@cilium.io> <aanm90@gmail.com>
<andre@cilium.io> <aanm@users.noreply.github.com>
```

After changing that, `git log --use-mailmap --format="%<|(40)%aN%aE" | sort -fu | grep Martins` only gives the single canonical email, and hence we'd get the right output in the AUTHORS file.

**Comment by shivam-Purohit** [comment]
@bimmlerd So we would need to fix the .mailmap manually right? something like this for Adam 
`Adam Bocim <adam.bocim@seznam.cz> <adam.bocim@firma.seznam.cz>`

so where should I get the data like through running for each author from git shorlog? like this
```
 git shortlog --summary \
		  | awk '{$1=""; print $0}' \
		  | sed -e 's/^ //' \
			-e '/vagrant/d'
```

**Comment by bimmlerd** [comment]
> @bimmlerd So we would need to fix the .mailmap manually right? 

Yeah, likely manually - it feels like a problem that doesn't suit automation all that well.

> something like this for Adam `Adam Bocim <adam.bocim@seznam.cz> <adam.bocim@firma.seznam.cz>`
> so where should I get the data?

Yeah, I'd take whatever is already in `AUTHORS` as the canonical name/email, and try to map other emails to that. These are not set in stone, it's about getting a good baseline without  duplication in the output of `extract_authors`. If necessary, authors can submit PRs to change the `.mailmap` themselves.

**Comment by shivam-Purohit** [comment]
> > @bimmlerd So we would need to fix the .mailmap manually right?
> 
> Yeah, likely manually - it feels like a problem that doesn't suit automation all that well.
> 
> > something like this for Adam `Adam Bocim <adam.bocim@seznam.cz> <adam.bocim@firma.seznam.cz>`
> > so where should I get the data?
> 
> Yeah, I'd take whatever is already in `AUTHORS` as the canonical name/email, and try to map other emails to that. These are not set in stone, it's about getting a good baseline without duplication in the output of `extract_authors`. If necessary, authors can submit PRs to change the `.mailmap` themselves.



> > @bimmlerd So we would need to fix the .mailmap manually right?
> 
> Yeah, likely manually - it feels like a problem that doesn't suit automation all that well.
> 
> > something like this for Adam `Adam Bocim <adam.bocim@seznam.cz> <adam.bocim@firma.seznam.cz>`
> > so where should I get the data?
> 
> Yeah, I'd take whatever is already in `AUTHORS` as the canonical name/email, and try to map other emails to that. These are not set in stone, it's about getting a good baseline without duplication in the output of `extract_authors`. If necessary, authors can submit PRs to change the `.mailmap` themselves.

Yeah i was thinking the same of using `AUTHORS `as .`mailmap `hasn't covered most of the authors. I will try to fix the mailmap as well as I can and will come up with a update.

**Comment by shivam-Purohit** [comment]
> Yeah, I'd take whatever is already in `AUTHORS` as the canonical name/email, and try to map other emails to that. These are not set in stone, it's about getting a good baseline without duplication in the output of `extract_authors`. If necessary, authors can submit PRs to change the `.mailmap` themselves.


Yeah i was thinking the same of using AUTHORS as .mailmap hasn't covered most of the authors. I will try to fix the mailmap as well as I can and will come up with a update.

**Comment by shivam-Purohit** [comment]
@bimmlerd I tired updating the mailmap using this python script
```
#!/usr/bin/env python3

import subprocess

import os
os.environ['LANG'] = 'en_US.UTF-8'
os.environ['LC_ALL'] = 'en_US.UTF-8'

def extract_authors():
    # Execute git command and capture output
    git_command = "git shortlog --summary | awk '{$1=\"\"; print $0}' | sed -e 's/^ //' -e '/vagrant/d'"
    authors_output = subprocess.check_output(git_command, shell=True, text=True)

    # Process the authors
    authors_list = authors_output.split('\n')

    authors_data = {}  # Dictionary to store authors' data

    for author in authors_list:
        # Get name
        name_command = f"git log --author='{author}' --format='%<|(40)%aN' | head -1"
        name = subprocess.check_output(name_command, shell=True, text=True).strip()

        # Get emails
        emails_command = f"git log --author='{author}' --format='%aE' | sort -u"
        emails = subprocess.check_output(emails_command, shell=True, text=True).strip().split('\n')

        # Update dictionary with author data
        if name in authors_data:
            authors_data[name].extend(emails)
        else:
            authors_data[name] = emails


    # Write results to file in the desired format with try-except block
    try:
        with open('testpythonscript.txt', 'w') as output_file:
            for name, emails in authors_data.items():
                output_file.write(f"{name} {' '.join(['<' + email + '>' for email in emails])}\n")

    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    extract_authors()

```

which can be reused. I copied the previous approach to map the author name to their emails. I thought of using map in python in which name can be set as key and email as an normal array. This should remove the repeating names in the mailmap but one problem is that some emails might repeat in the emails array. One solution to fix this can be using set instead of array so it can check for duplicate. But i removed ig all of the duplicate emails.

but STILL there are some problems when updating `AUTHORS`. Like this for Jed
```
Jed Salazar                              jed@isovalent.com
jedsalazar                               jed@isovalent.com
```
But as you can see i have set both the mails for jed in mailmap. When i search for` git log --author="jedsalazar"   `  I do get some entries, is that the problem here?
I will update the AUTHORS file to have some idea.

File: AUTHORS [hunk 1]
```diff
@@ -1,755 +1,774 @@
...
+Alex Szakaly                             alex.szakaly@gmail.com
+Alex Waring                              alex.waring@starlingbank.com
+Amey Bhide                               amey@covalent.io
+Amir Kheirkhahan                         amir.kheirkhahan@dbschenker.com
+amitmavgupta                             115551423+amitmavgupta@users.noreply.github.com
+Amol Ambekar                             ambekara@google.com
+Amre Shakimov                            amre@covalent.io
+Anderson, David L                        david.l.anderson@intel.com
+Andor Nemeth                             andor_nemeth@swissre.com
+AndrÃ© Martins                           andre@cilium.io
```
Comment (bimmlerd) [inline AUTHORS]: This is broken.

**Review by bimmlerd — CHANGES_REQUESTED** [review]
A few things:
1. The changes to the AUTHOR file seem wrong - there's too many changes, and I can see some malformed things sneaked in, such as in André Martins' case.
2. I don't think you will get around fixing up the `.mailmap` entries manually instead of through automation - I think e.g. spotting the wrong encoding of André's name is easier to do by hand than through automation.

Note that you don't need to add a `.mailmap` entry for each author - only for those who have used different names/emails and want to replace those with a single, canonical one. Maybe read through the mailmap docs again?

**Comment by shivam-Purohit** [comment]
> Note that you don't need to add a `.mailmap` entry for each author - only for those who have used different names/emails and want to replace those with a single, canonical one. Maybe read through the mailmap docs again?

yeah i got what mailmap does. sure then I will revert back the mailmap and fix only the broken ones ?

**Comment by shivam-Purohit** [comment]
@bimmlerd I think this should be fine. I updated the mailmap as suggested by you. I guess the repeated/duplicate values are not present. For some case like arthur it was also present in the original AUTHORS file. 
Using this script changes some emails as it uses different method to extract it. my commit was also added so the change of my name in AUTHORS files. I add added the AUTHORS file to have a reference.

File: .mailmap [hunk 2]
```diff
@@ -10,7 +11,7 @@ Arthur Chiao <arthurchiao@hotmail.com> <ArthurChiao@users.noreply.github.com>
 Arthur Evstifeev <aevstifeev@gitlab.com>
 Arthur Evstifeev <mail@ap4y.me>
 Arvind Soni <arvind@covalent.io>
-Ashwin Paranjpe <ashwin@covalent.io>
+Ashwin Paranjpe <ashwin@covalent.io>
```
Comment (bimmlerd) [inline .mailmap:14]: nit: excess space at the end?

File: .mailmap [hunk 3]
```diff
@@ -150,4 +157,4 @@ Yves Blusseau <yves.blusseau@acoss.fr>
 Xiaoqing <xiaoqingnb@gmail.com>
 Xiaoyang Zhu <zhuxiaoyang1996@gmail.com>
 Xin Li <xin.li@daocloud.io>
-Zhu Yan <hackzhuyan@gmail.com>
+Zhu Yan <hackzhuyan@gmail.com>
```
Comment (bimmlerd) [inline .mailmap:160]: nit: no newline at EOF

File: AUTHORS [hunk 4]
```diff
@@ -39,7 +39,7 @@ Andor Nemeth                            andor_nemeth@swissre.com
 Andreas Mårtensson                      andreas@addem.se
 Andree Klattenhoff                      mail@andr.ee
 Andrei Kvapil                           kvapss@gmail.com
-André Martins                           andre@cilium.io
+André Martins                           aanm90@gmail.com
```
Comment (bimmlerd) [inline AUTHORS:42]: I'm pretty sure that @aanm does not want his email to be changed to that gmail address. Why is this change here?

File: AUTHORS [hunk 5]
```diff
@@ -63,7 +63,7 @@ Antoine Legrand                         2t.antoine@gmail.com
 Antonio Ojea                            aojea@google.com
 Anton Ippolitov                         anton.ippolitov@datadoghq.com
 Antoni Zawodny                          zawodny@google.com
-Anton Protopopov                        aspsk@isovalent.com
+Anton Protopopov                        a.s.protopopov@gmail.com
```
Comment (bimmlerd) [inline AUTHORS:66]: Similarly, I also doubt that Anton wants this change.

**Review by bimmlerd — CHANGES_REQUESTED** [review]
The goal is to create a `.mailmap` file which causes the new command to not change the AUTHORS file. This isn't there yet.

Please also note that the changes should be contained either in a single commit, or two commits (one to change the mailmap, one to add the new command) but not more.

**Comment by shivam-Purohit** [comment]
> Please also note that the changes should be contained either in a single commit, or two commits (one to change the mailmap, one to add the new command) but not more.

i will quash the commit when done.

**Comment by shivam-Purohit** [comment]
> The goal is to create a `.mailmap` file which causes the new command to not change the AUTHORS file. This isn't there yet.

i will look into it.

**Comment by ti-mo** [comment]
Hi @shivam-Purohit, had another chance to look into this? I went ahead and closed https://github.com/cilium/cilium/pull/29936.

**Comment by shivam-Purohit** [comment]
@bimmlerd @ti-mo It's hard to get this through this method. This is sorting all the emails and picking the first. I cannot control which email will be picked through Mailmap. We will have to find some other way to look at this problem.

**Pull request closed without merging**